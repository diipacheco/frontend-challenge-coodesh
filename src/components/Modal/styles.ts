import styled from 'styled-components';

export const Content = styled.section`
  width: 100%;
  height: 100%;
  padding: 15px 30px;

  display: flex;
  align-items: center;
  flex-direction: column;

  position: relative;

  > h1 {
    margin-top: 25px;
  }

  div {
    margin-top: 14px;
  }
`;

export const UserInfoSection = styled.div`
  display: flex;

  width: 100%;
  align-self: flex-start;

  > strong {
    font-size: 18px;
    color: var(--blue);
  }

  > p {
    margin-left: 10px;
  }

  > a {
    font-size: 18px;
    color: var(--blue);
  }
`;

interface IAvatarContainer {
  urlAvatar: string;
}

export const AvatarContainer = styled.div<IAvatarContainer>`
  width: 100px;
  height: 100px;

  border-radius: 50%;

  background-image: ${props => `url(${props.urlAvatar})`};
  background-color: #ccc;
  background-repeat: no-repeat;
  background-size: contain;
`;

export const CloseModalButton = styled.button`
  position: absolute;
  cursor: pointer;

  top: -16px;
  right: -10px;
`;
