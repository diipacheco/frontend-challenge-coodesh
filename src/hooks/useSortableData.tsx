import { useState, useMemo, useCallback } from 'react';

interface ISortConfig {
  key: string;
  direction: string;
}

interface IUseSortableDataReturn {
  sortedItems: Array<any>;
  requestSort(key: string): void;
  sortConfig: ISortConfig | null;
}

export const useSortableData = (
  data: Array<any>,
  config = null,
): IUseSortableDataReturn => {
  const [sortConfig, setSortConfig] = useState<ISortConfig | null>(config);

  const sortedItems = useMemo(() => {
    const sortableItems = [...data];
    if (sortConfig) {
      sortableItems.sort((a, b) => {
        if (sortConfig.key === 'name') {
          if (a[sortConfig.key].first < b[sortConfig.key].first) {
            return sortConfig.direction === 'ascending' ? -1 : 1;
          }
          if (a[sortConfig.key].first > b[sortConfig.key].first) {
            return sortConfig.direction === 'ascending' ? 1 : -1;
          }
        }
        return 0;
      });
    }
    return sortableItems;
  }, [data, sortConfig]);

  const requestSort = useCallback(
    (key: string) => {
      let direction = 'ascending';
      if (
        sortConfig &&
        sortConfig?.key === key &&
        sortConfig?.direction === 'ascending'
      ) {
        direction = 'descending';
      }
      setSortConfig({ key, direction });
    },
    [sortConfig],
  );

  return { sortedItems, requestSort, sortConfig };
};
