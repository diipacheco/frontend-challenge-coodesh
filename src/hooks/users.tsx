import { createContext, useContext, useState, useCallback } from 'react';

import api from '../services/api';

interface ILocation {
  street: string;
  city: string;
  state: string;
  postcode: string;
}

export interface IUser {
  userId: number;
  id: {
    name: string;
    value: string;
  };
  name: {
    title: string;
    first: string;
    last: string;
  };
  picture: {
    large: string;
    medium: string;
    thumbnail: string;
  };
  email: string;
  gender: string;
  dob: {
    date: string;
    age: string;
  };
  cell: string;
  nat: string;
  location: ILocation;
}

interface IUserContextData {
  users: IUser[];
  user: IUser;

  fetchUsers(page: number): Promise<void>;
  filterUserInfo(id: string): void;
}

const UsersContext = createContext({} as IUserContextData);

export const UsersContextProvider: React.FC = ({ children }) => {
  const [users, setUsers] = useState([] as IUser[]);
  const [user, setUser] = useState({} as IUser);

  const fetchUsers = useCallback(async (page: number) => {
    const response = await api.get(
      `?inc=id,picture,name,email,gender,dob,cell,nat,location&results=50&page=${page}`,
    );
    setUsers(response.data.results);
  }, []);

  const filterUserInfo = useCallback(
    (id: string) => {
      const filteredUser = users?.find(item => item?.id?.value === id) as IUser;

      setUser(filteredUser);
    },
    [users],
  );

  return (
    <UsersContext.Provider value={{ users, user, fetchUsers, filterUserInfo }}>
      {children}
    </UsersContext.Provider>
  );
};

export function useUsers(): IUserContextData {
  const context = useContext(UsersContext);

  return context;
}
