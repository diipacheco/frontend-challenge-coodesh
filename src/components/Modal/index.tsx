import React from 'react';
import ReactModal from 'react-modal';
import moment from 'moment';
import { MdClose } from 'react-icons/md';
import { Link } from 'react-router-dom';

import { IUser } from '../../hooks/users';

import {
  Content,
  UserInfoSection,
  AvatarContainer,
  CloseModalButton,
} from './styles';

interface IModalProps {
  modalIsOpen: boolean;
  closeModal(): void;
  userInfo: IUser;
}

const customStyles = {
  overlay: {
    backgroundColor: 'rgb(2%, 2%, 2%, 0.2)',
  },

  content: {
    top: '50%',
    left: '50%',
    right: 'auto',
    bottom: 'auto',
    marginRight: '-50%',
    transform: 'translate(-50%, -50%)',
    width: 600,
    height: 530,
  },
};

const Modal: React.FC<IModalProps> = ({
  modalIsOpen,
  closeModal,
  userInfo,
}) => {
  return (
    <ReactModal
      style={customStyles}
      isOpen={modalIsOpen}
      ariaHideApp={false}
      onRequestClose={closeModal}
    >
      <Content>
        <CloseModalButton type="button" onClick={closeModal}>
          <MdClose size={30} />
        </CloseModalButton>
        <AvatarContainer urlAvatar={userInfo?.picture?.large} />

        <h1>{`${userInfo?.name?.first} ${userInfo?.name?.last}`}</h1>

        <UserInfoSection>
          <strong>Email:</strong>
          <p>{userInfo?.email}</p>
        </UserInfoSection>

        <UserInfoSection>
          <strong>Gender:</strong>
          <p>{userInfo?.gender}</p>
        </UserInfoSection>

        <UserInfoSection>
          <strong>Birth:</strong>
          <p>{moment(userInfo?.dob?.date).format('DD/MM/YYYY')}</p>
        </UserInfoSection>

        <UserInfoSection>
          <strong>Phone:</strong>
          <p>{userInfo?.cell}</p>
        </UserInfoSection>

        <UserInfoSection>
          <strong>NAT:</strong>
          <p>{userInfo?.nat}</p>
        </UserInfoSection>

        <UserInfoSection>
          <strong>Address:</strong>
          <p>{`${userInfo?.location?.street?.name}, ${userInfo?.location?.street?.number} - ${userInfo?.location?.state} - ${userInfo?.location?.city}`}</p>
        </UserInfoSection>

        <UserInfoSection>
          <strong>ID:</strong>
          <p>{`${userInfo?.id?.name} - ${userInfo?.id?.value}`}</p>
        </UserInfoSection>

        <UserInfoSection>
          <Link to={`/profile/${userInfo?.userId}`}>URL</Link>
        </UserInfoSection>
      </Content>
    </ReactModal>
  );
};

export default Modal;
