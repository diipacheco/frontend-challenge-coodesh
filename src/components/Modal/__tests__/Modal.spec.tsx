import { render } from '@testing-library/react';
import { Router } from 'react-router-dom';

import history from '../../../services/history';
import { IUser } from '../../../hooks/users';

import Modal from '../index';

describe('<Modal />', () => {
  it('should match snapshots', () => {
    const wrapper = render(
      <Router history={history}>
        <Modal modalIsOpen closeModal={jest.fn()} userInfo={{} as IUser} />
      </Router>,
    );

    expect(wrapper).toMatchSnapshot();
  });
});
