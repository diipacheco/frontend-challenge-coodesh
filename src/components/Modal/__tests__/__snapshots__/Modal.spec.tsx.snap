// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> should match snapshots 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="ReactModal__Body--open"
  >
    <div
      data-react-modal-body-trap=""
      style="position: absolute; opacity: 0;"
      tabindex="0"
    />
    <div />
    <div
      data-react-modal-body-trap=""
      style="position: absolute; opacity: 0;"
      tabindex="0"
    />
    <div
      class="ReactModalPortal"
    >
      <div
        class="ReactModal__Overlay"
        style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px;"
      >
        <div
          aria-modal="true"
          class="ReactModal__Content"
          role="dialog"
          style="position: absolute; top: 50%; left: 50%; border: 1px solid #ccc; background: rgb(255, 255, 255); overflow: auto; border-radius: 4px; outline: none; padding: 20px; margin-right: -50%; transform: translate(-50%, -50%); width: 600px; height: 530px;"
          tabindex="-1"
        >
          <section
            class="sc-bdfBQB gqMTeA"
          >
            <button
              class="sc-hKgJUU cTQnKd"
              type="button"
            >
              <svg
                fill="currentColor"
                height="30"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="30"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </button>
            <div
              class="sc-dlfnuX gNwICt"
            />
            <h1>
              undefined undefined
            </h1>
            <div
              class="sc-gsTEea fXhQjI"
            >
              <strong>
                Email:
              </strong>
              <p />
            </div>
            <div
              class="sc-gsTEea fXhQjI"
            >
              <strong>
                Gender:
              </strong>
              <p />
            </div>
            <div
              class="sc-gsTEea fXhQjI"
            >
              <strong>
                Birth:
              </strong>
              <p>
                07/10/2021
              </p>
            </div>
            <div
              class="sc-gsTEea fXhQjI"
            >
              <strong>
                Phone:
              </strong>
              <p />
            </div>
            <div
              class="sc-gsTEea fXhQjI"
            >
              <strong>
                NAT:
              </strong>
              <p />
            </div>
            <div
              class="sc-gsTEea fXhQjI"
            >
              <strong>
                Address:
              </strong>
              <p>
                undefined, undefined - undefined - undefined
              </p>
            </div>
            <div
              class="sc-gsTEea fXhQjI"
            >
              <strong>
                ID:
              </strong>
              <p>
                undefined - undefined
              </p>
            </div>
            <div
              class="sc-gsTEea fXhQjI"
            >
              <a
                href="/profile/undefined"
              >
                URL
              </a>
            </div>
          </section>
        </div>
      </div>
    </div>
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
